crates/matrix-sdk/src/notification_settings.rs

```diff
@@ -0,0 +1,789 @@
+        room_id: &String,
+        ruleset: &mut Ruleset,
+    ) -> Result<(), NotificationSettingsError> {
+        // remove any Override rules matching this room_id
+        for rule in ruleset.override_.clone().iter() {
+            if rule.rule_id == *room_id {
+                ruleset.remove(RuleKind::Override, &rule.rule_id)?;
+                continue;
+            }
+            if rule.conditions.iter().any(|x| match x {
+                PushCondition::EventMatch { key, pattern } => {
+                    key == "room_id" && *pattern == *room_id
+                }
+                _ => false,
+            }) {
+                ruleset.remove(RuleKind::Override, &rule.rule_id)?;
+                continue;
+            }
+        }
+
+        // remove any Room rules matching this room_id
+        if let Some(rule) = ruleset.room.clone().iter().find(|x| x.rule_id == *room_id) {
+            ruleset.remove(RuleKind::Room, &rule.rule_id)?;
+        }
+
+        // remove any Underride rules matching this room_id
+        for rule in ruleset.underride.clone().iter() {
+            if rule.rule_id == *room_id {
+                ruleset.remove(RuleKind::Underride, &rule.rule_id)?;
+                continue;
+            }
+            if rule.conditions.iter().any(|x| match x {
+                PushCondition::EventMatch { key, pattern } => {
+                    key == "room_id" && *pattern == *room_id
+                }
+                _ => false,
+            }) {
+                ruleset.remove(RuleKind::Underride, &rule.rule_id)?;
+                continue;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Gets the default push notification mode for a given room id
+    pub fn get_default_room_notification_mode(
+        &self,
+        is_encrypted: bool,
+        members_count: u64,
+        ruleset: &Ruleset,
+    ) -> Result<RoomNotificationMode, NotificationSettingsError> {
+        // get the correct default rule id based on is_encrypted and members_count
+        let rule_id = match (is_encrypted, members_count) {
+            (true, 2) => PredefinedUnderrideRuleId::EncryptedRoomOneToOne,
+            (false, 2) => PredefinedUnderrideRuleId::RoomOneToOne,
+            (true, _) => PredefinedUnderrideRuleId::Encrypted,
+            (false, _) => PredefinedUnderrideRuleId::Message,
+        };
+
+        // If there is an Underride rule that should trigger a notification, the mode is
+        // 'AllMessages'
+        if ruleset.underride.iter().any(|r| {
+            r.enabled
+                && r.rule_id == rule_id.to_string()
+                && r.actions.iter().any(|a| a.should_notify())
+        }) {
+            return Ok(RoomNotificationMode::AllMessages);
+        }
+
+        // If user mention is enabled, the mode is 'MentionsAndKeywordsOnly'
+        if self.is_user_mention_enabled(ruleset)
+            || self.contains_keyword_rules(ruleset)
+            || self.is_room_mention_enabled(ruleset)
```

Comment:
@nimau we decided to remove this sanity check from the SDK. We will return M&K by default. We will let the application handle this border use case where all mentions are turned off